circuits/sybil/sybilUtils.circom

```diff
@@ -0,0 +1,104 @@
+
+template ComputeSybilID(){
+    signal input stateSecret;
+    signal input claimHash;
+    signal input crs;
+
+    signal output out;
+
+    component hash = Poseidon(3);
+    hash.inputs[0] <== stateSecret;
+    hash.inputs[1] <== claimHash;
+    hash.inputs[2] <== crs;
+    out <== hash.out;
+}
+
+// constants
+template GetStateSecretIndex() {
+    signal output out;
+    out <== 0xbb67ae85;
+}
+
+template GetStateSecretSchemaHash(){
+    signal output schemaHash;
+    schemaHash <== 301485908906857522017021291028488077057;
+}
+
+
+template VerifyAndExtractValStateSecret(holderLevels, gistLevels){
+    signal input claim[8];
+    signal input claimIssuanceMtp[holderLevels];
+    signal input claimIssuanceClaimsRoot;
+    signal input claimIssuanceRevRoot;
+    signal input claimIssuanceRootsRoot;
+    signal input claimIssuanceIdenState; // aka idenGistState
+
+    // signal input claimSchema;
+
+    signal input idenGistMtp[gistLevels];
+    // signal input idenGistState;              // double check this declration 
+    signal input gistRoot;
+
+    signal input userId;
+
+    // inter
+    signal expectedClaimIdx;
+    signal stateSecretSchema;
+
+    signal output secret;
+    // (1) Verify claim is included in claims tree root
+    component claimIssuanceCheck = checkClaimExists(holderLevels);
+    for (var i=0; i<8; i++) { claimIssuanceCheck.claim[i] <== claim[i]; }
+    for (var i=0; i<holderLevels; i++) { claimIssuanceCheck.claimMTP[i] <== claimIssuanceMtp[i]; }
+    claimIssuanceCheck.treeRoot <== claimIssuanceClaimsRoot;
+
+    // verify state includes claims tree
+    component verifyClaimIssuanceIdenState = checkIdenStateMatchesRoots();
+    verifyClaimIssuanceIdenState.claimsTreeRoot <== claimIssuanceClaimsRoot;
+    verifyClaimIssuanceIdenState.revTreeRoot <== claimIssuanceRevRoot;
+    verifyClaimIssuanceIdenState.rootsTreeRoot <== claimIssuanceRootsRoot;
+    verifyClaimIssuanceIdenState.expectedState <== claimIssuanceIdenState;
+
+    component stateSecretSchemaHash = GetStateSecretSchemaHash();
+    stateSecretSchema <== stateSecretSchema.schemaHash;
+
+    // (2) Verify claim schema
+    component claimSchemaCheck = verifyCredentialSchema();
+    for (var i=0; i<8; i++) { claimSchemaCheck.claim[i] <== claim[i]; }
+    claimSchemaCheck.schema <== stateSecretSchema;
+
+    // (3) Verify issuer state is in GIST 
+    component claimIssuanceCheckGISTInc = checkClaimExists(gistLevels);
+    // claimIssuanceIdenState
+    // for (var i=0; i<8; i++) { claimIssuanceCheckGISTInc.claim[i] <== claimIssuanceIdenState[i]; }
+    for (var i=0; i<gistLevels; i++) { claimIssuanceCheckGISTInc.claimMTP[i] <== idenGistMtp[i]; }
+    claimIssuanceCheckGISTInc.treeRoot <== gistRoot;
```

Comment:
Why do you need to check by the checkClaimExists template here? You don't work with the claims in the GIST tree. This tree contains **hash(id)** as indexes and **states** as values in their leaves, so checking that is enough

Reply:
You are looking on an old commit, this file has been updated since then (after the conversation we had with Dima), here is the current version - https://github.com/iden3/circuits/blob/8a5ce88710422450fc1a536f0fa36dbfcea3a0f1/circuits/sybil/sybilUtils.circom#L78 with the SMT verifier